PyPI Openzwave

JshWright opened the pull request:
## Description:

Initial PR to switch to PyPI based openzwave installation. Documentation PR forthcoming shortly.

**Related issue (if applicable):** fixes #7403 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<Coming Soon>

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

JshWright committed:
Remove default zwave config path

PYOZW now has much more comprehensive default handling for the config
path (in src-lib/libopenzwave/libopenzwave.pyx:getConfig()). It looks in
the same place we were looking, plus _many_ more. It will certainly do a
much better job of finding the config files than we will (and will be
updated as the library is changed, so we don't end up chasing it). The
getConfig() method has been there for a while, but was subsntially
improved recently.

This change simply leaves the config_path as None if it is not
specified, which will trigger the default handling in PYOZW.

JshWright committed:
Install python-openzwave from PyPI

As of version 0.4, python-openzwave supports installation from PyPI,
which means we can use our 'normal' dependency management tooling to
install it. Yay.

This uses the default 'embed' build (which goes and downloads
statically sources to avoid having to compile anything locally). Check
out the python-openzwave readme for more details.

balloob commented on homeassistant/components/zwave/__init__.py:
```diff
@@ -221,22 +221,12 @@ def setup(hass, config):
     from openzwave.network import ZWaveNetwork
     from openzwave.group import ZWaveGroup
 
-    default_zwave_config_path = os.path.join(os.path.dirname(
```
Is the configuration not a part of the package? That is unfortunate. I wonder if we should auto download the config in a future PR. Let's make sure that we note in the docs how to obtain it.

JshWright replied:
No, it does, it's just in a different place. The behavior when config_path is None is now basically an improved version of what we were doing (it checks the new location, along with a number of other places where it has been historically). There's no longer a need for us to get clever with the default path.

JshWright commented:
Not sure what's up with Travis. Tests run fine in a clean virtualenv locally. Will have to take a look tomorrow.

balloob commented:
```
    g++ -pthread -shared -L/opt/python/3.4.2/lib -Wl,-rpath=/opt/python/3.4.2/lib build/temp.linux-x86_64-3.4/openzwave-embed/open-zwave-master/python-openzwave/src-lib/libopenzwave/libopenzwave.o openzwave-embed/open-zwave-master/libopenzwave.a -L/opt/python/3.4.2/lib -ludev -lstdc++ -lresolv -lpython3.4m -o build/lib.linux-x86_64-3.4/libopenzwave.cpython-34m.so
    g++: error: openzwave-embed/open-zwave-master/libopenzwave.a: No such file or directory
    error: command 'g++' failed with exit status 1
```

eperdeme commented:
If this gets merged its also worth at the same time removing; 

virtualization/Docker/scripts/python_openzwave

As it won't be needed and;

virtualization/Docker/setup_docker_prereqs

and;
virtualization/Docker/setup_docker_prereqs
@@ -30,7 +30,8 @@ PACKAGES=(
 PACKAGES_DEV=(
-  cython3 libudev-dev
+  libudev-dev

 )

eperdeme commented:
I think the travis errors actually due to;

   b'/tmp/pip-build-jwlooorh/python-openzwave/openzwave-embed/open-zwave-master/cpp/hidapi/linux/hid.c:46:21: fatal error: libudev.h: No such file or directory\n'
    b'/tmp/pip-build-jwlooorh/python-openzwave/openzwave-embed/open-zwave-master/cpp/hidapi/linux/hid.c:46:21: fatal error: libudev.h: No such file or directory\n'\

You need libudev deb pkg installed to compile it.

balloob committed:
Add python-openzwave deps to .travis.yml

Python OpenZwave require the libudev headers to build. This adds the
libudev-dev package to Travis runs via the 'apt' addon for Travis.

Thanks to @MartinHjelmare for this fix.

balloob commented:
@eperdeme good catch. I added `apt-get install libudev-dev` to our travis.yml, see if this can get it building.

balloob commented:
This is getting messy. Now we need to apt-get install a bunch of stuff.

Since we don't need this dependency for tests, we should consider to rollback all changes to travis.yml and instead make it one of the requirements that we comment out: https://github.com/home-assistant/home-assistant/blob/dev/script/gen_requirements_all.py#L9

The only downside to that is that it will also make sure that it won't get installed in the Docker container, although I guess we can address that.

andrey-git commented:
Whats the downside of `apt-get install` ?
Although we don't need it for tests - it follows more closely the environment the users need to have.

MartinHjelmare commented:
Container based env on travis (`sudo: false`) has much faster boot time. Using the APT `addon` is preferred over `sudo apt-get install`. Then you can keep `sudo: false` if setting `system_site_packages: true` under `virtualenv`. Eg:

```yml
virtualenv:
  system_site_packages: true
sudo: false
addons:
  apt:
    packages:
    - python-numpy
```

MartinHjelmare commented:
Got this build time the second time after a caching run:
https://travis-ci.org/MartinHjelmare/home-assistant/builds/229042786

JshWright commented:
Thanks for the .travis.yml tweaks @MartinHjelmare. I've force pushed an update to this branch including those tweaks (I ended up adding a new commit, as you commit had a couple other things included in it).

JshWright committed:
Update docker build for PyPI openzwave

Now that PYOZW can be install from PyPI, the docker image build process
can be simplified to remove the explicit compilation of PYOZW.

andrey-git commented:
I think the docker should keep `libudev-dev

balloob commented:
@andrey-git it has been added to the main install script.

I think that this is ok to merge. The 0.44 branch has been cut so it means that people will be able to test this for two weeks.

This PR has slowed down the tests (as Travis now has to compile OpenZWave). I am planning to update `gen_requirements_all.py` to make a specific test file that just installs the dependencies necessary for tests. That way we should get a pretty good boost.

hawk259 commented:
Any idea on how to toggle this and build your own version OZW and PYOZW in the docker image?

I copy/tweaked the old python_openzwave and call it in the docker build, but that didn't work.  Also tried to:

`pip3 uninstall -y python_openzwave`

and then build/install python_openzwave and that didn't work.

I next attempt will be to patch requirements_all.txt and homeassistant/components/zwave/__init__.py and see if that might be it (remove the requirement).

The reason I need this is I need to patch OZW and PYOZW to add support for the [Linear NGDZ00-4 Garage Door](https://community.home-assistant.io/t/linear-ngdz00-4-garage-door/2136)

balloob commented:
Home Assistant will not install the python Open Z-Wave dependency if it has already been installed in the current Python environment. Make sure the version number is the same though.

hawk259 commented:
Just following up:

So it looks like you have to git clone the exact version of python-openzwave from requirements_all.txt otherwise it will uninstall it and re-install the version the requirements_all.txt version.

It seems python-openzwave 0.4.x changed the way it builds and installs.  A build target will git clone OZW in the openzwave directory and build it.  A install target will download a OZW source tar file, expand it to openzwave-embed, build and install it with the rest of python-openzwave.

Once I figured that out and patched both copies of OZW, I have a working zwave garage door cover again.

I am testing the 3.6 image change now and will also test just doing a install instead of build/install.
